runtime/compiler/trj9/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -5735,36 +5735,24 @@ static void genHeapAlloc(TR::Node *node, TR::Instruction *&iCursor, TR_OpaqueCla
-               if (elementSize >= 4)
+               if (elementSize >= 2)
                   {
                   if (TR::Compiler->target.is64Bit())
-                     iCursor = generateShiftLeftImmediateLong(cg, node, dataSizeReg, enumReg, trailingZeroes(elementSize), iCursor);
+                     iCursor = generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldic, node, dataSizeReg, enumReg, trailingZeroes(elementSize), CONSTANT64(0x00000000FFFFFFFF) << trailingZeroes(elementSize)); 
                   else
                      iCursor = generateShiftLeftImmediate(cg, node, dataSizeReg, enumReg, trailingZeroes(elementSize), iCursor);
 
                   }
-               else if (elementSize == 2)
+               if (needZeroInit && elementSize <= 2)
                   {
-                  // RS64 processors prefer adds to shifts
-                  iCursor = generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, dataSizeReg, enumReg, enumReg, iCursor);
-                  if (needZeroInit)
-                     {
-                     // the zero initialization code uses a loop of stwu's, and
-                     // so dataSizeReg must be rounded up to a multiple of 4
+                  // the zero initialization code uses a loop of stwu's, and
+                  // so dataSizeReg must be rounded up to a multiple of 4
+                  if (elementSize == 2) 
                      iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, dataSizeReg, dataSizeReg, 3, iCursor);
-                     if (TR::Compiler->target.is64Bit())
-                        iCursor = generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicr, node, dataSizeReg, dataSizeReg, 0, int64_t(-4), iCursor);
-                     else
-                        iCursor = generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, dataSizeReg, dataSizeReg, 0, -4, iCursor);
-                     }
-                  }
-               else if (needZeroInit) // elementSize == 1
-                  {
-                  // dataSizeReg must be set in this case for use by the zero
-                  // initializing code, and it must be rounded up to a multiple
-                  // of 4 (for the loop of stwu's)
-                  iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, dataSizeReg, enumReg, 3, iCursor);
-                  if (TR::Compiler->target.is64Bit())
+                  else  //elementSize == 1
+                     iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, dataSizeReg, enumReg, 3, iCursor);
+
+                  if (TR::Compiler->target.is64Bit() && elementSize != 1)
                      iCursor = generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicr, node, dataSizeReg, dataSizeReg, 0, int64_t(-4), iCursor);
                   else
                      iCursor = generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, dataSizeReg, dataSizeReg, 0, -4, iCursor);
```

Comment:
This existing (2nd) rounding of size looked unnecessary to me, although I knew fixing this is not part of the goal/mandate of this PR.  It has no bad side-effect when round is bigger than 4, and that is true anyway.